src/main/java/frc/robot/subsystems/elevator/ElevatorIOReal.java

```diff
@@ -1,15 +1,96 @@
 package frc.robot.subsystems.elevator;
 
+import static edu.wpi.first.units.Units.Degrees;
+import static edu.wpi.first.units.Units.Meters;
+
+import com.ctre.phoenix6.controls.DutyCycleOut;
+import com.ctre.phoenix6.controls.MotionMagicVoltage;
+import com.ctre.phoenix6.controls.StrictFollower;
+import com.ctre.phoenix6.hardware.TalonFX;
 import edu.wpi.first.units.Distance;
 import edu.wpi.first.units.MutableMeasure;
+import edu.wpi.first.units.Units;
+import edu.wpi.first.wpilibj.DigitalInput;
+import edu.wpi.first.wpilibj.Servo;
+import frc.robot.Ports;
 
 public class ElevatorIOReal implements ElevatorIO {
+    private final Servo servo;
+    private final TalonFX mainMotor;
+    private final TalonFX auxMotor;
+
+    private final DigitalInput bottomSensor;
+    private final DigitalInput topSensor;
+
+    private final MotionMagicVoltage positionControl = new MotionMagicVoltage(0);
+    private final DutyCycleOut powerControl = new DutyCycleOut(0);
+
+    public ElevatorIOReal() {
+        mainMotor = new TalonFX(Ports.Elevator.MAIN_MOTOR);
+        auxMotor = new TalonFX(Ports.Elevator.AUX_MOTOR);
+        servo = new Servo(Ports.Elevator.SERVO);
+        bottomSensor = new DigitalInput(Ports.Elevator.BOTTOM_LIMIT_SWITCH);
+        topSensor = new DigitalInput(Ports.Elevator.TOP_LIMIT_SWITCH);
+
+        mainMotor.getConfigurator().apply(ElevatorConstants.MOTOR_CONFIGURATION);
+        auxMotor.getConfigurator().apply(ElevatorConstants.MOTOR_CONFIGURATION);
+
+        auxMotor.setControl(new StrictFollower(mainMotor.getDeviceID()));
+    }
+
     @Override
-    public void setPower(double power) {}
+    public void setPower(double power) {
+        mainMotor.setControl(powerControl.withOutput(power));
+    }
 
     @Override
-    public void setHeight(MutableMeasure<Distance> height) {}
+    public void setHeight(MutableMeasure<Distance> height) {
+        inputs.heightSetpoint.mut_replace(height);
+        mainMotor.setControl(positionControl.withPosition(height.in(Units.Meters)));
+    }
 
     @Override
-    public void updateInputs(ElevatorInputs inputs) {}
+    public void updateInputs(ElevatorInputs inputs) {
+        inputs.isBottom = atBottom();
+        inputs.isTop = atTop();
+        inputs.carriageHeight = Meters.of(mainMotor.getPosition().getValue()).mutableCopy();
+
+        inputs.gripperHeight.mut_replace(
+                inputs.carriageHeight.gt(ElevatorConstants.GRIPPER_HEIGHT)
+                        ? inputs.carriageHeight
+                                .mutableCopy()
+                                .mut_minus(ElevatorConstants.GRIPPER_HEIGHT)
+                        : Meters.zero());
+
+        inputs.servoAngle = Units.Degrees.of(servo.getAngle()).mutableCopy();
+    }
+
+    public boolean atTop() {
+        return topSensor.get();
+    }
+
+    public boolean atBottom() {
+        return bottomSensor.get();
+    }
```

Comment:
What are the usages of these getters?

Reply:
To know when the elevator at bottom and reset the encoder accordingly

Reply:
As stated above, this can be handled by the Talon FX itself.

Reply:
36b85ab1ea79db7c196561265572b3c460aa3bae

Reply:
Nice!
Note that there are two kinds of faults. Did you use the correct type? You can read about the differences in the documentation.

Reply:
Changed to Fault